### PR TITLE
WIP: counter: add callbacks for Add, Delete events

### PR DESCRIPTION
**Pull request by ianvernon**
Add ways to provide callbacks upon add and delete events if necessary for the
prefix counter. This allows for factoring out of datapath dependencies out of
policy creation / deletion functionality.

Signed-off by: Ian Vernon <ian@cilium.io>

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/9442)

### CODE REVIEW

**Comment by ianvernon** [comment]
test-me-please

File: daemon/daemon.go [hunk 1]
```diff
@@ -553,6 +555,15 @@ func NewDaemon(dp datapath.Datapath, iptablesManager rulesManager) (*Daemon, *en
 	return &d, restoredEndpoints, nil
 }
 
+func (d *Daemon) OnPrefixCountChange() error {
+	if !ipcachemap.BackedByLPM() {
+		// Only recompile if configuration has changed.
+		log.Debug("CIDR policy has changed; recompiling base programs")
```
Comment (joestringer) [inline daemon/daemon.go:561]: Apparently we had this as debug before, but me might consider logging this at a higher level like `Info`. It's pretty invasive and knowing the time that this happened would be useful if anything happened to go wrong.

File: pkg/counter/prefixes.go [hunk 2]
```diff
@@ -32,6 +32,13 @@ type PrefixLengthCounter struct {
 
 	maxUniquePrefixes4 int
 	maxUniquePrefixes6 int
+
+	listenerLock lock.RWMutex
```
Comment (joestringer) [inline pkg/counter/prefixes.go:36]: Any particular reason this needs its own lock?

File: pkg/counter/prefixes.go [hunk 3]
```diff
@@ -45,6 +52,15 @@ func NewPrefixLengthCounter(maxUniquePrefixes6, maxUniquePrefixes4 int) *PrefixL
 	}
 }
 
+func (p *PrefixLengthCounter) AddListener(listener prefixCountChangeListener) {
+	p.listenerLock.Lock()
+	defer p.listenerLock.Unlock()
+	if p.listeners == nil {
```
Comment (joestringer) [inline pkg/counter/prefixes.go:58]: If this nil-check is only for handling the initialization case, then we should probably just do this step in `NewPrefixLengthCounter()`.

**Review by joestringer — COMMENTED** [review]
Overall this is good in that it hides some detail away from daemon and the policy management code, but at the same time since it hides that away, it feels like it's more likely to get us into trouble because we forget this occurs and it's hidden behind layers.

That said I don't have another proposal which achieves your aims so LGTM modulo the minor nits below.

**Pull request closed without merging**